### PR TITLE
Change auto-update logic to match semver

### LINKED ISSUE #343: Update to ClassicPress definition of "minor" and "major" versions
The WordPress update system considers an update from `1.0` to `1.1` as a **major** version upgrade, which will not be applied automatically under the default configuration.

ClassicPress sites can safely apply these upgrades automatically, because `1.0.0` -> `1.1.0` will be a minor version bump, which is [guaranteed](https://www.classicpress.net/roadmap/) not to contain any breaking changes.

So, in the ClassicPress code, we need to change the definition of a **major** version upgrade:

- `1.0.0` -> `1.1.0` is currently considered a **major** version upgrade, but it should be considered as a **minor** version update instead
- `1.0.0` -> `2.0.0` should remain flagged as a **major** version upgrade
- `1.0.0` -> `1.0.1` should remain flagged as a **minor** version upgrade (under semver rules it is actually a **patch** version upgrade, but we probably won't introduce that distinction right now)

More information about relevant constants/filters [here](https://make.wordpress.org/core/2013/10/25/the-definitive-guide-to-disabling-auto-updates-in-wordpress-3-7/).

### PR DESCRIPTION
**Pull request by invisnet**
Code changes to fix #343.

* ~Removes `allow_dev_auto_core_updates` filter~
* Adds `allow_nightly_auto_core_updates` filter
* Adds `allow_patch_auto_core_updates`filter

We need to update some docs to reflect these changes - I'm not sure where best to do that.

~There are no tests for this - I'm out of time this week, I can pick this up again on Monday or someone else is welcome to jump in.~

### CODE REVIEW

**Comment by nylen** [comment]
Thanks for the PR.  Next steps:

- Use `$cp_version` instead of `$wp_version`
- Split this function into 2 so that it can be tested
- Add automated tests
- Do some manual tests of automatic updates (using custom builds of CP with a custom API URL or some other hack to fake the availability of new versions)

**Comment by invisnet** [comment]
> Thanks for the PR. Next steps:
> 
> * Use `$cp_version` instead of `$wp_version`

Ha! That's a classic wood-for-the-trees one that is.

> * Split this function into 2 so that it can be tested

Where are you thinking of splitting it? I don't see an obvious place where splitting it into 2 would help - I can see how 4 would be better, but not 2.

File: src/wp-admin/includes/class-core-upgrader.php [hunk 1]
```diff
@@ -292,38 +292,14 @@ public function upgrade( $current, $args = array() ) {
 	 * @return bool True if we should update to the offered version, otherwise false.
 	 */
 	public static function should_update_to_version( $offered_ver ) {
-		include( ABSPATH . WPINC . '/version.php' ); // $wp_version; // x.y.z
```
Comment (nylen) [inline src/wp-admin/includes/class-core-upgrader.php]: The split to make this logic testable is here after this first line.

- `should_update_to_version()` just collects data - runs `include`, packages up the relevant constants into an array, and passes off to another function
- `auto_update_enabled_for_versions()` contains the core logic as a pure function (inputs map directly to outputs without depending on any other external state), and this is the function that's tested in unit tests

File: src/wp-admin/includes/class-core-upgrader.php [hunk 2]
```diff
@@ -333,34 +309,80 @@ public static function should_update_to_version( $offered_ver ) {
 				return false;
 
 			// Don't claim we can update on update-core.php if we have a non-critical failure logged.
-			if ( $wp_version == $failure_data['current'] && false !== strpos( $offered_ver, '.1.next.minor' ) )
+			if ( $cp_version == $failure_data['current'] && false !== strpos( $offered_ver, '.1.next.minor' ) )
```
Comment (nylen) [inline src/wp-admin/includes/class-core-upgrader.php]: @invisnet do you know what this `.1.next.minor` text is?  Where does this come from, and do we need to do something else instead?

File: src/wp-admin/includes/class-core-upgrader.php [hunk 3]
```diff
@@ -333,34 +309,80 @@ public static function should_update_to_version( $offered_ver ) {
...
+			if ( empty( $failure_data['retry'] ) && $cp_version == $failure_data['current'] && $offered_ver == $failure_data['attempted'] )
 				return false;
 		}
 
-		// 3: 3.7-alpha-25000 -> 3.7-alpha-25678 -> 3.7-beta1 -> 3.7-beta2
-		if ( $current_is_development_version ) {
+		// that concludes all the sanity checks, now we can do some work
+
+		$ver_current = preg_split( '/[.-\+]/', $cp_version  );
+		$ver_offered = preg_split( '/[.-\+]/', $offered_ver );
```
Comment (nylen) [inline src/wp-admin/includes/class-core-upgrader.php]: `$ver_offered = $offered_ver` :upside_down_face: I'm going to change this.

File: src/wp-admin/includes/class-core-upgrader.php (hunk 2 above)
Comment (invisnet) [reply src/wp-admin/includes/class-core-upgrader.php]: I have no idea so I left it as-is. It's something to track down for testing.

Comment (nylen) [reply src/wp-admin/includes/class-core-upgrader.php]: Found it - it is precisely what it says in the comment `// Don't claim we can update on update-core.php if we have a non-critical failure logged.`

The rest of the related code is here:

https://github.com/ClassicPress/ClassicPress/blob/8797d5eaf7a06de6814fd91d2d6c25fd8ff27bdb/src/wp-admin/update-core.php#L188-L199

**Pull request merged**